File: packages/js/src/where.ts
import { CrudianError, assertString } from "./errors.js"

export type Op =
  | "eq"
  | "ne"
  | "lt"
  | "gt"
  | "lte"
  | "gte"
  | "in"
  | "like"
  | "isNull"
  | "isNotNull"

export type CondNode = {
  type: "cond"
  op: Op
  column: string
  value?: unknown
}

export type GroupNode = {
  type: "and" | "or"
  children: WhereNode[]
}

export type WhereNode = CondNode | GroupNode

function cond(op: Op, column: string, value?: unknown): CondNode {
  assertString(column, "column")
  return { type: "cond", op, column, value }
}

/** Public where builder. Tree shape is an internal representation via `toNode()`. */
export class WhereBuilder {
  private readonly node: WhereNode

  private constructor(node: WhereNode) {
    this.node = node
  }

  static create(): WhereBuilder {
    return new WhereBuilder({ type: "and", children: [] })
  }

  static from(node: WhereNode): WhereBuilder {
    return new WhereBuilder(node)
  }

  private appendCond(op: Op, column: string, value?: unknown): WhereBuilder {
    const next = cond(op, column, value)
    if (this.node.type === "and") {
      return new WhereBuilder({ type: "and", children: [...this.node.children, next] })
    }
    return new WhereBuilder({ type: "and", children: [this.node, next] })
  }

  eq(column: string, value: unknown): WhereBuilder {
    return this.appendCond("eq", column, value)
  }

  ne(column: string, value: unknown): WhereBuilder {
    return this.appendCond("ne", column, value)
  }

  lt(column: string, value: unknown): WhereBuilder {
    return this.appendCond("lt", column, value)
  }

  gt(column: string, value: unknown): WhereBuilder {
    return this.appendCond("gt", column, value)
  }

  lte(column: string, value: unknown): WhereBuilder {
    return this.appendCond("lte", column, value)
  }

  gte(column: string, value: unknown): WhereBuilder {
    return this.appendCond("gte", column, value)
  }

  in(column: string, value: unknown[]): WhereBuilder {
    if (!Array.isArray(value)) {
      throw new CrudianError("in value must be an array")
    }
    return this.appendCond("in", column, value)
  }

  like(column: string, value: unknown): WhereBuilder {
    return this.appendCond("like", column, value)
  }

  isNull(column: string): WhereBuilder {
    return this.appendCond("isNull", column)
  }

  isNotNull(column: string): WhereBuilder {
    return this.appendCond("isNotNull", column)
  }

  and(...others: WhereBuilder[]): WhereBuilder {
    const children = [this.toNode(), ...others.map((b) => b.toNode())]
    return new WhereBuilder({ type: "and", children })
  }

  or(...others: WhereBuilder[]): WhereBuilder {
    const children = [this.toNode(), ...others.map((b) => b.toNode())]
    return new WhereBuilder({ type: "or", children })
  }

  toNode(): WhereNode {
    return this.node
  }
}

export function where(): WhereBuilder {
  return WhereBuilder.create()
}

export function isWhereBuilder(value: unknown): value is WhereBuilder {
  return value instanceof WhereBuilder
}

File: packages/go/crudian/where.go
package crudian

type Op string

const (
	OpEq        Op = "eq"
	OpNe        Op = "ne"
	OpLt        Op = "lt"
	OpGt        Op = "gt"
	OpLte       Op = "lte"
	OpGte       Op = "gte"
	OpIn        Op = "in"
	OpLike      Op = "like"
	OpIsNull    Op = "isNull"
	OpIsNotNull Op = "isNotNull"
)

type CondNode struct {
	Type   string
	Op     Op
	Column string
	Value  any
}

type GroupNode struct {
	Type     string // and | or
	Children []WhereNode
}

// WhereNode is CondNode or GroupNode.
type WhereNode interface {
	whereNode()
}

func (CondNode) whereNode()  {}
func (GroupNode) whereNode() {}

// WhereBuilder is the public fluent API.
type WhereBuilder struct {
	node WhereNode
}

func Where() *WhereBuilder {
	return &WhereBuilder{node: GroupNode{Type: "and", Children: nil}}
}

func (w *WhereBuilder) ToNode() WhereNode { return w.node }

func (w *WhereBuilder) appendCond(op Op, column string, value any) *WhereBuilder {
	next := CondNode{Type: "cond", Op: op, Column: column, Value: value}
	if g, ok := w.node.(GroupNode); ok && g.Type == "and" {
		ch := append(append([]WhereNode{}, g.Children...), next)
		return &WhereBuilder{node: GroupNode{Type: "and", Children: ch}}
	}
	return &WhereBuilder{node: GroupNode{Type: "and", Children: []WhereNode{w.node, next}}}
}

func (w *WhereBuilder) Eq(column string, value any) *WhereBuilder {
	return w.appendCond(OpEq, column, value)
}
func (w *WhereBuilder) Ne(column string, value any) *WhereBuilder {
	return w.appendCond(OpNe, column, value)
}
func (w *WhereBuilder) Lt(column string, value any) *WhereBuilder {
	return w.appendCond(OpLt, column, value)
}
func (w *WhereBuilder) Gt(column string, value any) *WhereBuilder {
	return w.appendCond(OpGt, column, value)
}
func (w *WhereBuilder) Lte(column string, value any) *WhereBuilder {
	return w.appendCond(OpLte, column, value)
}
func (w *WhereBuilder) Gte(column string, value any) *WhereBuilder {
	return w.appendCond(OpGte, column, value)
}
func (w *WhereBuilder) In(column string, value any) *WhereBuilder {
	return w.appendCond(OpIn, column, value)
}
func (w *WhereBuilder) Like(column string, value any) *WhereBuilder {
	return w.appendCond(OpLike, column, value)
}
func (w *WhereBuilder) IsNull(column string) *WhereBuilder {
	return w.appendCond(OpIsNull, column, nil)
}
func (w *WhereBuilder) IsNotNull(column string) *WhereBuilder {
	return w.appendCond(OpIsNotNull, column, nil)
}

func (w *WhereBuilder) And(other *WhereBuilder) *WhereBuilder {
	if other == nil {
		return w
	}
	return &WhereBuilder{node: GroupNode{Type: "and", Children: []WhereNode{w.node, other.node}}}
}

func (w *WhereBuilder) Or(other *WhereBuilder) *WhereBuilder {
	if other == nil {
		return w
	}
	return &WhereBuilder{node: GroupNode{Type: "or", Children: []WhereNode{w.node, other.node}}}
}
